presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/cache/ReentrantExecutor.java

```diff
@@ -0,0 +1,43 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.prestosql.plugin.hive.metastore.cache;
+
+import java.util.concurrent.Executor;
+
+import static java.util.Objects.requireNonNull;
+
+public class ReentrantExecutor
+        implements Executor
+{
+    private final ThreadLocal<Boolean> executorThreadMarkers = ThreadLocal.withInitial(() -> false);
+    private final Executor delegate;
+
+    public ReentrantExecutor(Executor delegate)
+    {
+        this.delegate = requireNonNull(delegate, "delegate is null");
+    }
+
+    @Override
+    public void execute(Runnable task)
+    {
+        if (executorThreadMarkers.get()) {
+            task.run();
```

Comment:
Previously the call would run on a separate thread. Now it runs on the stack.
It would be better to extended BoundedExecutor, or alternatively, share the underlying executor and delegate to that.

Reply:
then the class would be `ReentrantBoundedExecutor`